diff -up usbguard-0.7.0/configure.ac.libaudit-version usbguard-0.7.0/configure.ac
--- usbguard-0.7.0/configure.ac.libaudit-version	2017-12-13 10:31:18.738471922 +0100
+++ usbguard-0.7.0/configure.ac	2017-12-13 10:32:10.763256142 +0100
@@ -230,7 +230,7 @@ AC_SUBST([crypto_LIBS])
 #
 # Check for optional libraries
 #
-PKG_CHECK_MODULES([audit], [audit >= 2.7.7],
+PKG_CHECK_MODULES([audit], [audit >= 2.7.0],
 [AC_DEFINE([HAVE_LINUX_AUDIT], [1], [Linux Audit API available])
 AC_CHECK_DECLS([audit_encode_nv_string], [], [], [[#include<libaudit.h>]])
 libaudit_summary="system-wide; $audit_LIBS"],
