diff --git a/src/Library/DeviceManagerPrivate.cpp b/src/Library/DeviceManagerPrivate.cpp
index d9f3395..579b3c7 100644
--- a/src/Library/DeviceManagerPrivate.cpp
+++ b/src/Library/DeviceManagerPrivate.cpp
@@ -41,6 +41,7 @@ namespace usbguard {
     std::unique_lock<std::mutex> local_device_map_lock(_device_map_mutex);
     std::unique_lock<std::mutex> remote_device_map_lock(rhs._device_map_mutex);
     _device_map = rhs._device_map;
+    _restore_controller_device_state = rhs._restore_controller_device_state;
     return *this;
   }
 
diff --git a/src/Library/Exception.hpp b/src/Library/Exception.hpp
index 02e9622..f27dd63 100644
--- a/src/Library/Exception.hpp
+++ b/src/Library/Exception.hpp
@@ -128,7 +128,8 @@ namespace usbguard
   {
     public:
       IPCException()
-        : Exception("", "", "")
+        : Exception("", "", ""),
+	  _message_id(0)
       {
       }
 
